docs/cmd/kn_service_migrate.md

```diff
@@ -0,0 +1,53 @@
+## kn service migrate
+
+Migrate Knative services from source cluster to destination cluster
+
+### Synopsis
+
+Migrate Knative services from source cluster to destination cluster
+
+```
+kn service migrate [flags]
+```
+
+### Examples
+
+```
+
+  # Migrate Knative services from source cluster to destination cluster by export KUBECONFIG as environment variables
+  kn migrate --source-namespace default --destination-namespace default
+
+  # Migrate Knative services from source cluster to destination cluster by set kubeconfig as parameters
+  kn migrate --source-namespace default --destination-namespace default --source-kubeconfig $HOME/.kube/config/source-cluster-config.yml --destination-kubeconfig $HOME/.kube/config/destination-cluster-config.yml
+
+  # Migrate Knative services from source cluster to destination cluster and force replace the service if exists in destination cluster
+  kn migrate --source-namespace default --destination-namespace default --force
+
+  # Migrate Knative services from source cluster to destination cluster and delete the service in source cluster
+  kn migrate --source-namespace default --destination-namespace default --force --delete
+```
+
+### Options
+
+```
+      --delete                          Delete all Knative resources after kn-migration from source cluster
+      --destination-kubeconfig string   The kubeconfig of the destination Knative resources (default is KUBECONFIG2 from ENV property)
```

Comment:
Why would one prefer to different kubeconfig files instead of two different Kubernetes contexts ?. I.e. a kubeconfig file can contain many contexts, and the source and destination should map directly to a context not to kubeconfig's default context only.

I.e. you could also want to migrate from one to another context stored in the _same_ kubeconfig file. It's not the file which defines a Kubernetes cluster (but the context within the config file)

Reply:
My suggestion would be:

* Allow multiple kubeconfig files like we already do by supporting a path like structure for the env var (`KUBECONFIG=firstconfig:secondconfig`)
* For specifying source and target, we by default both the default context, but one can override this with `--source-context` and `--target-context`, respectively.

That way we would stay consistent with the current handling of the kubeconfi configuration.

Reply:
Hi @rhuss ,

I think if it is a migration case, people won't like to save two kubeconfig in one file.

In most of the cases, after migration, the source cluster may be removed or never be used.

So I prefer to use `--kubeconfig` as other commands for source cluster and use `--destination-kubeconfig` for target cluster. then user doesn't need to copy kube confi to the one file manually.

Pls let me know if you have other comment, thanks a lot!

Reply:
I concur with rhuss's suggestion. Should we open a knob for migrating ksvcs using cluster context?

> user doesn't need to copy kube confi to the one file manually.

You can use a single kubeconfig file for managing multiple clusters regardless of the provider. So you just add new context for the destination cluster and delete it after migration if you would like to do by using kubectl or other CLI tools. Some CLI tools (`gcloud`, `eksctl`, ...) automatically add a context to `~/.kube/config` file (or where you want to put it in) and set it to current context when creating a new cluster.

1. An cluster admin creates first K8S cluster and saves its context in `~/.kube/config` file with `first-cluster`

1. An cluster admin creates a ksvc in first K8S cluster using current context (`first-cluster`)

    ```bash
    kn service create hello --image "gcr.io/knative-samples/helloworld-go"
    ```

1. An cluster admin creates second K8S cluster and saves its context in `~/.kube/config` file with `second-cluster`

1. An cluster admin migrates existing ksvc into new K8S cluster

    ```bash
    # If a current context is set to target cluster:
    kn service migrate hello --source-context first-cluster
    # or a current context is set to source cluster:
    kn service migrate hello --target-context second-cluster
    ```

1. An cluster admin delete source context from `~/.kube/config` file

    ```bash
    kubectl config delete-cluster <first cluster name>
    kubectl config delete-cluster first-cluster
    kubectl config  unset users.<user name>
    ```